SYCL/ESIMD/api/functional/ctors/ctor_broadcast.hpp

```diff
@@ -0,0 +1,210 @@
+    const auto simd_by_init = esimd::simd<DstT, NumElems>(ref_value);
+    simd_by_init.copy_to(out);
+  }
+};
+
+// Descriptor class for the case of calling constructor in variable declaration
+// context.
+struct var_decl {
+  static std::string get_description() { return "variable declaration"; }
+
+  template <typename SrcT, typename DstT, int NumElems>
+  static void call_simd_ctor(SrcT ref_value, DstT *const out) {
+    const esimd::simd<DstT, NumElems> simd_by_var_decl(ref_value);
+    simd_by_var_decl.copy_to(out);
+  }
+};
+
+// Descriptor class for the case of calling constructor in rvalue in an
+// expression context.
+struct rval_in_expr {
+  static std::string get_description() { return "rvalue in an expression"; }
+
+  template <typename SrcT, typename DstT, int NumElems>
+  static void call_simd_ctor(SrcT ref_value, DstT *const out) {
+    esimd::simd<DstT, NumElems> simd_by_rval;
+    simd_by_rval = esimd::simd<DstT, NumElems>(ref_value);
+    simd_by_rval.copy_to(out);
+  }
+};
+
+// Descriptor class for the case of calling constructor in const reference
+// context.
+class const_ref {
+public:
+  static std::string get_description() { return "const reference"; }
+
+  template <typename SrcT, typename DstT, int NumElems>
+  static void call_simd_ctor(SrcT ref_value, DstT *const out) {
+    call_simd_by_const_ref<SrcT, DstT, NumElems>(
+        esimd::simd<DstT, NumElems>(ref_value), out);
+  }
+
+private:
+  template <typename SrcT, typename DstT, int NumElems>
+  static void
+  call_simd_by_const_ref(const esimd::simd<DstT, NumElems> &simd_by_const_ref,
+                         DstT *const out) {
+    simd_by_const_ref.copy_to(out);
+  }
+};
+
+template <typename SrcT, typename DstT, int NumElems, typename ContextT>
+class BroadcastCtorTestDescription : public ITestDescription {
+public:
+  BroadcastCtorTestDescription(size_t index, DstT retrieved_val,
+                               DstT expected_val, SrcT ref_value,
+                               const std::string &src_data_type,
+                               const std::string &dst_data_type)
+      : m_src_data_type(src_data_type), m_dst_data_type(dst_data_type),
+        m_retrieved_val(retrieved_val), m_expected_val(expected_val),
+        m_ref_val(ref_value), m_index(index) {}
+
+  std::string to_string() const override {
+    // TODO: Make strings for fp values more short during failure output, may be
+    // by using hex representation
+    std::string log_msg("Failed for simd<");
+
+    log_msg += m_src_data_type + ", " + std::to_string(NumElems) + ">";
+    log_msg += ", with context: " + ContextT::get_description();
+    log_msg += ", with source value type: " + m_src_data_type;
+    log_msg += ", with destination value type: " + m_dst_data_type;
+    log_msg += ", retrieved: " + std::to_string(m_retrieved_val);
+    log_msg += ", expected: " + std::to_string(m_expected_val);
+    log_msg += ", ref value: " + std::to_string(m_ref_val);
```

Comment:
```suggestion
    log_msg += ", source type: " + m_src_data_type;
    log_msg += ", destination type: " + m_dst_data_type;
    log_msg += ", retrieved: " + std::to_string(m_retrieved_val);
    log_msg += ", expected: " + std::to_string(m_expected_val);
    log_msg += ", input: " + std::to_string(m_ref_val);
```

Also, as a suggestion.
So, currently we have the following message, for example:
`Failed for simd<float,8>, with context: variable declaration, with source value type: float, with destination value type: int, retrieved: 42, expected: 41, ref value: 41.5, at index: 0`

What if we have something like this instead:
`Failed for simd<float,8>, with context: variable declaration, destination type: int. Unexpected value at index 0. Input: 41.5, retrieved: 42, expected: 41`

Reply:
Whoops, one more thing. I see the following code:
```
  template <typename SrcT, typename DstT, int NumElems>
  static void call_simd_ctor(SrcT ref_value, DstT *const out) {
    const auto simd_by_init = esimd::simd<DstT, NumElems>(ref_value);
    simd_by_init.copy_to(out);
  }
```
Is using `"Failed for simd<" +  m_src_data_type + ", " + std::to_string(NumElems) + ">"` valid here?

Reply:
> What if we have something like this instead:

No, I'd like to provide source and destination types

Reply:
this comment applied in this commit: da1b71a5bd6811a44a5ea48d03bb911518fe1655

Reply:
Shouldn't we use `"Failed for simd<" + m_dst_data_type`?

Reply:
this comment was applied [here](https://github.com/intel/llvm-test-suite/pull/690/commits/be308ab37a2211afe7f091450cf5fb97c3e6346e#diff-056299db9fa515e78e72e7973d8a03b15b2a0ddb3557d03188e25a7b9b6edae1L97-R97)